### PR TITLE
BL-4344 Allow hint bubbles per field

### PR DESCRIPTION
**Pull request by JohnThomson**
This change is [Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/1561)

### CODE REVIEW

**Comment by JohnThomson** [comment]
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomBrowserUI/bookEdit/TextBoxProperties/TextBoxProperties.ts, line 425 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1561#-Kf80KJzTuuyOufWpXgB:-Kf80KJzTuuyOufWpXgC:bz5a6t3) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1d4e24474e9148fafeed91a591b1225ed8e11e6c/src/BloomBrowserUI/bookEdit/TextBoxProperties/TextBoxProperties.ts#L425)):*
> ```XML
>     }
> 
>     showHintClassName(): string { return "bloom-showHintOnEach" }
> ```

This name deserves special attention as it gets saved in the Bloom HTML files so we can't easily change it once this version ships

---

*[src/BloomExe/web/EnhancedImageServer.cs, line 397 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1561#-Kf80Vmfj9E7pGTaFqyY:-Kf80VmgbobSqM1vF3Gh:b48ayim) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1d4e24474e9148fafeed91a591b1225ed8e11e6c/src/BloomExe/web/EnhancedImageServer.cs#L397)):*
> ```Smalltalk
> 						langs.Add(new {label=WorkspaceView.MenuItemName(lang), tag=lang.IetfLanguageTag});
> 					}
> 					info.ContentType = "application/json";
> ```

Not really part of this change but it fixes one JS error that was coming up while launching the dialog

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1561)*

File: src/BloomBrowserUI/bookEdit/TextBoxProperties/TextBoxProperties.less [hunk 1]
```diff
@@ -13,11 +22,17 @@
 // select2 select controls are way too narrow if we don't force a width.
 // Need a considerable one for some of the longer language names.
 #lang-select + .select2-container{
-    width: 250px !important;
+    width: @lang-select-width !important;
     margin-top: 10px;
     margin-left: @hint-tab-indent;
 }
 
+#show-when + .select2-container{
```
Comment (hatton) [inline src/BloomBrowserUI/bookEdit/TextBoxProperties/TextBoxProperties.less]: In the broader context here, this name is inscrutable. How about hint-scope?

**Comment by JohnThomson** [comment]
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/BloomBrowserUI/bookEdit/TextBoxProperties/TextBoxProperties.less, line 30 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1561#-Kf808zfYLkQeSQYYAq_-r2-30:-KfSRYi3YPGFeptj49MD:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/5dee92f628650a06e3bcaf1b8ae0b7d6327df41b/src/BloomBrowserUI/bookEdit/TextBoxProperties/TextBoxProperties.less#L30)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

In the broader context here, this name is inscrutable. How about hint-scope?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1561)*

**Comment by andrew-polk** [comment]
Reviewed 4 of 6 files at r1.
Review status: 2 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/BloomBrowserUI/bookEdit/TextBoxProperties/TextBoxProperties.ts, line 425 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1561#-Kf80KJzTuuyOufWpXgB:-Kf81u1KYrt20gWxthwb:bdsra8c) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1d4e24474e9148fafeed91a591b1225ed8e11e6c/src/BloomBrowserUI/bookEdit/TextBoxProperties/TextBoxProperties.ts#L425)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

This name deserves special attention as it gets saved in the Bloom HTML files so we can't easily change it once this version ships
</blockquote></details>

I was confused by this name as I thought a call to the function would show something. 
Not sure how to improve it, though... classNameForShowingHints?

---

*[src/BloomBrowserUI/bookEdit/TextBoxProperties/TextBoxProperties.ts, line 436 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1561#-Kf82NDkP1PfD14whF-G:-Kf82NDkP1PfD14whF-H:bbp4awy) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1d4e24474e9148fafeed91a591b1225ed8e11e6c/src/BloomBrowserUI/bookEdit/TextBoxProperties/TextBoxProperties.ts#L436)):*
> languageGroupChecked = $('input[name="languageRadioGroup"]:checked').val() == "Auto"

This logic seems backwards to me. Isn't a language group checked if it isn't Auto?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1561)*

**Comment by JohnThomson** [comment]
Review status: 2 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/BloomBrowserUI/bookEdit/TextBoxProperties/TextBoxProperties.ts, line 425 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1561#-Kf80KJzTuuyOufWpXgB:-KfSVwgvy8KrGjveIb38:b-g9l21q) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1d4e24474e9148fafeed91a591b1225ed8e11e6c/src/BloomBrowserUI/bookEdit/TextBoxProperties/TextBoxProperties.ts#L425)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I was confused by this name as I thought a call to the function would show something. 
Not sure how to improve it, though... classNameForShowingHints?
</blockquote></details>

changed to classNameForHintOnEach()

---

*[src/BloomBrowserUI/bookEdit/TextBoxProperties/TextBoxProperties.ts, line 436 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1561#-Kf82NDkP1PfD14whF-G:-KfSWt7PqNSkax5hCXoi:bf9krci) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1d4e24474e9148fafeed91a591b1225ed8e11e6c/src/BloomBrowserUI/bookEdit/TextBoxProperties/TextBoxProperties.ts#L436)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> languageGroupChecked = $('input[name="languageRadioGroup"]:checked').val() == "Auto"

This logic seems backwards to me. Isn't a language group checked if it isn't Auto?
</blockquote></details>

"Auto" is the value associated with the radio button (on the first page) which causes the program to show whatever languages the book is configured for. The other radio buttons on that page each select exactly one language, making this "show on each" control irrelevant.
I've changed the name to groupCanHaveMoreThanOneLanguage in the hope of making things clearer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1561)*

**Pull request merged**